### PR TITLE
Feature security group provider

### PR DESCRIPTION
**Pull request by varikin**

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        }
+      }
+    }
+
+    securityGroups
+  }
+
+  @Override
+  String getType() {
+    return OpenstackCloudProvider.ID
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Remove return keyword.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAllByAccountAndRegion(boolean includeRule, String account, String region) {
+    getAllByAccount(includeRule, account).findAll { it.region == region }
+  }
+
+  @Override
+  OpenstackSecurityGroup get(String account, String region, String name, String vpcId) {
+    getAllByAccountAndRegion(true, account, region).findAll { it.name == name }[0]
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: The findResult method will iterate through the collection and find the first non-null result.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  Set<OpenstackSecurityGroup> getSecurityGroups() {
+    // TODO This is not thread safe and not cached
+    // This just proves the provider is working once it has a data set.
+    Set<OpenstackSecurityGroup> securityGroups = []
+    accountCredentialsProvider.all.each {account ->
+      if (account instanceof OpenstackNamedAccountCredentials) {
+        def provider = ((OpenstackNamedAccountCredentials) account).credentials.provider
+
+        provider.allRegions.each { region ->
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: This seems a bit clumsy ... Here is how I would do it in Java8 streams api.

`regions.stream()
    .map{ r -> provider.getSecurityGroups(r) }
    .flatMap{ sg -> this.&mapToSecurityGroup(sg) }
    .collect(Collectors.toCollection(() -> securityGroups))`

Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: I agree, though this is just a place holder till the security groups caching is in place. Then this provider will just grab the data from the cache.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 2 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Yes. That is the intent.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  Set<OpenstackSecurityGroup> getSecurityGroups(boolean includeRules) {
+    // TODO This is not thread safe and not cached
+    // This just proves the provider is working once it has a data set.
+    Set<OpenstackSecurityGroup> securityGroups = []
+    accountCredentialsProvider.all.each { account ->
+      if (account instanceof OpenstackNamedAccountCredentials) {
+        def provider = ((OpenstackNamedAccountCredentials) account).credentials.provider
+
+        provider.allRegions.each { region ->
+          securityGroups.addAll(provider.collect { p ->
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Don't need to collect on a single object.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackNamedAccountCredentials
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider
+import groovy.util.logging.Slf4j
+import org.openstack4j.model.compute.SecGroupExtension
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+@Slf4j
+@Component
+class OpenstackSecurityGroupProvider implements SecurityGroupProvider<OpenstackSecurityGroup> {
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Add a TODO in here to replace direct lookup with cache lookup, and then implement security group caching agents.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 2 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Can you change to use findResult?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 4 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: Hmmm..not sure what I was thinking...

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 2 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy]: I change to `find`. I thought I tried both of these first and it didn't handle the case of no matches. Works this time, don't know what I did wrong before ¯_(ツ)_/¯

**Pull request closed without merging**